src/DataTypes/getLeastSupertype.h

```diff
@@ -48,4 +50,24 @@ DataTypePtr getLeastSupertypeOrString(const TypeIndexSet & types);
 
 DataTypePtr tryGetLeastSupertype(const TypeIndexSet & types);
 
+/// A map that enumerated all interval kinds in ascending order with a conversion value to a next interval
+inline const std::unordered_map<IntervalKind::Kind, std::pair<int, int>> & getGranularityMap()
```

Comment:
Actually it might be better with a single vector, (maybe some inlined vector, but I don't think it is necessary) where the values are ordered by `IntervalKind`. My reasons are:
1. In `findSmallestIntervalSuperType` iterating over a small vector should be fast enough
2. in the conversion function you could just iterate over a range of this vector to find the conversion factor instead of using searching the right value all the time.

Using a regular map is also viable, but because it is a small constant map, we can do log(n) serach in the same way if not faster (because of cache locality).